File: dotnet/src/EspBarcode.Protocol/Cobs.cs
namespace EspBarcode.Protocol;

public static class Cobs
{
    public static byte[] Encode(ReadOnlySpan<byte> data)
    {
        var output = new List<byte>(data.Length + data.Length / 254 + 2);
        int codeIndex = 0;
        output.Add(0); // placeholder for the first code byte, backpatched below
        byte code = 1;
        for (int i = 0; i < data.Length; i++)
        {
            if (data[i] == 0x00)
            {
                output[codeIndex] = code;
                codeIndex = output.Count;
                output.Add(0); // placeholder for the next code byte
                code = 1;
            }
            else
            {
                output.Add(data[i]);
                code++;
                if (code == 0xFF)
                {
                    output[codeIndex] = code;
                    codeIndex = output.Count;
                    output.Add(0); // placeholder
                    code = 1;
                }
            }
        }
        output[codeIndex] = code; // finalize the last (possibly empty) group unconditionally
        return [.. output];
    }

    public static bool TryDecode(ReadOnlySpan<byte> data, out byte[] decoded)
    {
        var output = new List<byte>(data.Length);
        int read = 0;
        while (read < data.Length)
        {
            byte code = data[read];
            if (code == 0) { decoded = []; return false; }
            read++;
            int blockLen = code - 1;
            if (read + blockLen > data.Length) { decoded = []; return false; }
            for (int i = 0; i < blockLen; i++) output.Add(data[read + i]);
            read += blockLen;
            if (code != 255 && read < data.Length) output.Add(0x00);
        }
        decoded = [.. output];
        return true;
    }
}

File: lib/EspLinkCore/src/Cobs.cpp
#include "Cobs.h"

namespace esplink {

std::vector<uint8_t> cobsEncode(const uint8_t* data, std::size_t length) {
    std::vector<uint8_t> out;
    out.reserve(length + length / 254 + 2);
    std::size_t codeIndex = 0;
    out.push_back(0);  // placeholder for the first code byte, backpatched below
    uint8_t code = 1;
    for (std::size_t i = 0; i < length; ++i) {
        if (data[i] == 0x00) {
            out[codeIndex] = code;
            codeIndex = out.size();
            out.push_back(0);  // placeholder for the next code byte
            code = 1;
        } else {
            out.push_back(data[i]);
            ++code;
            if (code == 0xFF) {
                out[codeIndex] = code;
                codeIndex = out.size();
                out.push_back(0);  // placeholder
                code = 1;
            }
        }
    }
    out[codeIndex] = code;  // finalize the last (possibly empty) group unconditionally
    return out;
}

bool cobsDecode(const uint8_t* data, std::size_t length, std::vector<uint8_t>& out) {
    out.clear();
    std::size_t read = 0;
    while (read < length) {
        const uint8_t code = data[read];
        if (code == 0) return false;  // a literal zero byte is never valid inside a COBS block.
        ++read;
        const std::size_t blockLen = std::size_t(code) - 1;
        if (read + blockLen > length) return false;
        out.insert(out.end(), data + read, data + read + blockLen);
        read += blockLen;
        if (code != 255 && read < length) out.push_back(0x00);
    }
    return true;
}

}  // namespace esplink
